src/driver/safe/external_memory.rs

```diff
@@ -0,0 +1,124 @@
+use core::{marker::PhantomData, mem::ManuallyDrop};
+use std::fs::File;
+use std::ops::Range;
+use std::sync::Arc;
+
+use super::{CudaDevice, DevicePtr, DeviceSlice};
+use crate::driver::{result, sys, DriverError};
+
+impl CudaDevice {
+    /// Import external memory from a [`File`].
+    ///
+    /// # Safety
+    /// `size` must be the size of the external memory.
+    #[cfg(any(unix, windows))]
+    pub unsafe fn import_external_memory(
+        self: &Arc<Self>,
+        file: File,
+        size: u64,
+    ) -> Result<ExternalMemory, DriverError> {
+        #[cfg(unix)]
+        let external_memory = unsafe {
+            use std::os::fd::AsRawFd;
+            result::external_memory::import_external_memory_opaque_fd(file.as_raw_fd(), size)
+        }?;
+        #[cfg(windows)]
+        let external_memory = unsafe {
+            use std::os::windows::io::AsRawHandle;
+            result::external_memory::import_external_memory_opaque_win32(file.as_raw_handle(), size)
+        }?;
+        Ok(ExternalMemory {
+            external_memory,
+            size,
+            _device: self.clone(),
+            _file: ManuallyDrop::new(file),
+        })
+    }
+}
+
+/// An abstraction for imported external memory.
+///
+/// This struct can be created via [`CudaDevice::import_external_memory`].
+/// The imported external memory will be destroyed when this struct is dropped.
+#[derive(Debug)]
+pub struct ExternalMemory {
+    external_memory: sys::CUexternalMemory,
+    size: u64,
+    _device: Arc<CudaDevice>,
+    _file: ManuallyDrop<File>,
```

Comment:
I'm not familiar with ManuallyDrop, what's the benefit of wrapping this?

Reply:
[`ManuallyDrop`](https://doc.rust-lang.org/stable/std/mem/struct.ManuallyDrop.html) allows you to avoid calling `drop` of the inner type when it goes out of scope. Ideally I would just use [`std::mem::forget`](https://doc.rust-lang.org/std/mem/fn.forget.html) in the `Drop` implementation, but I can't because I cannot move an attribute out while I have a mutable reference. I also cannot do the usual trick of [`std::mem::take`](https://doc.rust-lang.org/std/mem/fn.take.html) because `File` does not implement `Default`.

The reason we need to avoid dropping is because on Unix, the CUDA driver takes ownership of the file handle and any operations on it after are UB. On Windows though, you have to drop the handle yourself. See the [cuda docs here](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__EXTRES__INTEROP.html#group__CUDA__EXTRES__INTEROP_1g52aba3a7f780157d8ba12972b2481735) (scroll down until you get to "If [CUDA_EXTERNAL_MEMORY_HANDLE_DESC::type](https://docs.nvidia.com/cuda/cuda-driver-api/structCUDA__EXTERNAL__MEMORY__HANDLE__DESC__v1.html#structCUDA__EXTERNAL__MEMORY__HANDLE__DESC__v1_1d4e3663348d28278d066980b422ab70e) is [CU_EXTERNAL_MEMORY_HANDLE_TYPE_OPAQUE_FD](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1ggbca0bdde9a33c47058b5c97f21e2edd864efb8402f2268e489336396b2048c07) ...").

If you prefer to avoid `ManuallyDrop` I could also use an `Option` and that would allow me to use `std::mem::take` and `std::mem::forget`.

Reply:
Oh cool. I like ManuallyDrop for this, nice usage!